contrib/long_distance_matching/circular_buffer_table.c

```diff
@@ -0,0 +1,107 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "ldm_hashtable.h"
+#include "mem.h"
+
+//TODO: move def somewhere else.
+//TODO: memory usage is currently no longer LDM_MEMORY_USAGE.
+//      refactor code to scale the number of elements appropriately.
+
+// Number of elements per hash bucket.
+#define HASH_BUCKET_SIZE_LOG 0 // MAX is 4 for now
+#define HASH_BUCKET_SIZE (1 << (HASH_BUCKET_SIZE_LOG))
+
+struct LDM_hashTable {
+  U32 size;
+  LDM_hashEntry *entries;  // 1-D array for now.
+
+  // Position corresponding to offset=0 in LDM_hashEntry.
+  const BYTE *offsetBase;
+  BYTE *bucketOffsets;     // Pointer to current insert position.
+
+                           // Last insert was at bucketOffsets - 1?
+};
+
+LDM_hashTable *HASH_createTable(U32 size, const BYTE *offsetBase) {
+  LDM_hashTable *table = malloc(sizeof(LDM_hashTable));
+  table->size = size;
+  table->entries = calloc(size * HASH_BUCKET_SIZE, sizeof(LDM_hashEntry));
+  table->bucketOffsets = calloc(size, sizeof(BYTE));
+  table->offsetBase = offsetBase;
+  return table;
+}
+
+static LDM_hashEntry *getBucket(const LDM_hashTable *table, const hash_t hash) {
+  return table->entries + (hash << HASH_BUCKET_SIZE_LOG);
+}
+
+LDM_hashEntry *HASH_getValidEntry(const LDM_hashTable *table,
```

Comment:
Here are some stats from compressing the file. The ldm example was run with
`#define LDM_MIN_MATCH_LENGTH 64`
`#define HASH_BUCKET_SIZE_LOG 3`
`#define LDM_WINDOW_SIZE_LOG 28`
`#define LDM_MEMORY_USAGE 23`

^ I made some changes so memory usage of 2^23 and bucket size of 2^3 implies 2^16 buckets. 

File size: 568645152

Method |  Size | Ratio
-------- | -----| -----
lrzip -p 32 -n -w 1        |	      181887800	|      3.13
ldm 			|			      153828356	|      3.70
lz4                                   | 	      104075804	|      5.46
lrzip -p 32 -l -w 1          |           	       88496017	|      6.43
lrzip -p 32 -n                 |         	       80074779	|      7.10
zstd -1                          |       	       69203650	|      8.22
lrzip -p 32 -n -w 1 & zstd -1          | 	       66116676	 |     8.60
gzip -9                           |      	       64613027 |	      8.80
lrzip -p 32 -g -L 9 -w 1        |        	       58757173	|      9.68
zstd -T16 -9                           | 	       54564744	|     10.42
lrzip -p 32 -n -w 1 & zstd -T32 -19   |  	       49499696	|     11.49
zstd -T32 -19                         |  	       47919263	|     11.87
ldm & zstd -1 				|	       26363651   |   21.57	
lrzip -p 32 -n & zstd -1             |   	       21549633	|     26.39
lrzip -p 32 -g -L 9                    | 	       19035126	   |  29.87
lrzip -o 32 -n & zstd -T32 -19     |    	       16350887	 |    34.78